### PR TITLE
Switch from the old node runner to the new job runner

### PR DESCRIPTION
**Pull request by chenhao94**

### CODE REVIEW

**Comment by chenhao94** [comment]
Ping

File: src/message.cc [hunk 1]
```diff
@@ -47,15 +48,9 @@ void CRMessageBase::Dispatch(const CRMessageBasePtr& message) {
...
-        if (!queue) [[unlikely]] {
-            LOG(ERROR) << __func__ << ": node: " << node << ", no queue for message " << message->GetMessageTypeName()
-                       << ", skipping";
-            continue;
-        }
-        queue->AddMessage(CRMessageBasePtr(message));
-        node->Kick();
+    for (auto& node : subscription_info->sub_list_) {
+        [[maybe_unused]] bool success = node->AddMessageToRunner(std::move(message));
+        DCHECK(success);
```
Comment (zhouxindong) [inline src/message.cc:53]: `DCHECK(node->AddMessageToRunner(std::move(message)))` 这样是不是就可以不需要[[maybe_unused]]了？

Comment (chenhao94) [reply src/message.cc:53]: 不行，dcheck在非debug模式下不存在，那这样你addmessage也没执行

File: src/node.cc [hunk 2]
```diff
@@ -0,0 +1,71 @@
+#include "cris/core/node.h"
+
+#include "cris/core/logging.h"
+
+namespace cris::core {
+
+CRNode::~CRNode() {
+    for (auto&& subscribed : subscribed_) {
+        CRMessageBase::Unsubscribe(subscribed, this);
```
Comment (zhouxindong) [inline src/node.cc:9]: 反正已经是析构了 std::move(subscribed)是不是可以节约一点？

**Review by zhouxindong — APPROVED** [review]

**Comment by chenhao94** [comment]
@xkszltl any comments?

File: includes/node.h [hunk 3]
```diff
@@ -19,73 +21,54 @@ class CRNode {
...
-    void WaitForMessage(duration_t&& timeout);
+    bool AddJobToRunner(job_t&& job);
+
+    bool AddMessageToRunner(const CRMessageBasePtr& message);
 
     // Not thread-safe, do not call concurrently nor call it
-    // when messages are coming, nor call it after Node Runner
-    // is Running
+    // when messages are coming, nor call it after runner
+    // is running
```
[the remark this answers is not in the document]
Comment (xkszltl) [reply includes/node.h:48]: And why not align to the 120 limit?

File: src/message.cc [hunk 4]
```diff
@@ -47,15 +48,9 @@ void CRMessageBase::Dispatch(const CRMessageBasePtr& message) {
...
-        auto* queue = node->MessageQueueMapper(message);
-        if (!queue) [[unlikely]] {
-            LOG(ERROR) << __func__ << ": node: " << node << ", no queue for message " << message->GetMessageTypeName()
-                       << ", skipping";
-            continue;
-        }
-        queue->AddMessage(CRMessageBasePtr(message));
-        node->Kick();
+    for (auto& node : subscription_info->sub_list_) {
+        [[maybe_unused]] bool success = node->AddMessageToRunner(std::move(message));
```
Comment (xkszltl) [inline src/message.cc:52]: const

File: src/node.cc [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    }
+}
+
+bool CRNode::AddMessageToRunner(const CRMessageBasePtr& message) {
+    if (auto job_opt = GenerateJob(message)) {
+        return AddJobToRunner(std::move(*job_opt));
```
Comment (xkszltl) [inline src/node.cc:15]: include

File: src/node.cc [hunk 6]
```diff
@@ -0,0 +1,71 @@
+#include "cris/core/node.h"
+
+#include "cris/core/logging.h"
+
+namespace cris::core {
+
+CRNode::~CRNode() {
+    for (auto&& subscribed : subscribed_) {
```
Comment (xkszltl) [inline src/node.cc:8]: auto&

File: src/node.cc [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+std::optional<CRNode::msg_callback_t> CRNode::GetCallback(const CRMessageBasePtr& message) {
+    if (!message) {
+        return std::nullopt;
+    }
+
+    const auto channel = message->GetChannelId();
+
+    auto callback_search_result = callbacks_.find(channel);
```
Comment (xkszltl) [inline src/node.cc:40]: const?

File: tests/node_test.cc [hunk 8]
```diff
@@ -2,133 +2,250 @@
...
-        threads.emplace_back([&unblocked, &node]() {
-            node.WaitForMessage(std::chrono::seconds(1));
-            unblocked.fetch_add(1);
+        std::mutex cv_mutex;
+        auto       cv       = std::make_shared<std::condition_variable>();
+        auto       received = std::make_shared<std::atomic<bool>>(false);
+
+        const int message_value = 0;
+
+        // Different type, never called
```
Comment (xkszltl) [inline tests/node_test.cc:41]: .

File: tests/node_test.cc [hunk 9]
```diff
@@ -2,133 +2,250 @@
...
+        subscriber.Subscribe<TestMessage<10086>>(channel_subid, [](const auto&) {
+            // Never enters.
+            std::abort();
+        });
+
+        // Different subid, never called
+        subscriber.Subscribe<TestMessageType>(10010, [](const auto&) {
```
Comment (xkszltl) [inline tests/node_test.cc:48]: .

File: tests/node_test.cc [hunk 10]
```diff
@@ -2,133 +2,250 @@
...
+
+        subscriber.Subscribe<TestMessageType>(
+            channel_subid,
+            [message_value, cv, received](const std::shared_ptr<TestMessageType>& msg) {
+                EXPECT_EQ(msg->value_, message_value);
+                received->store(true);
+                cv->notify_all();
+            });
+
+        // duplicate subscription, first one wins
```
Comment (xkszltl) [inline tests/node_test.cc:61]: cap and .

File: tests/node_test.cc [hunk 11]
```diff
@@ -2,133 +2,250 @@
...
+        auto   runner = std::make_unique<JobRunner>(JobRunner::Config{});
+        subscriber.SetRunner(runner.get());
 
-    for (std::size_t i = 0; i < kThreadNum; ++i) {
-        threads.emplace_back([&unblocked, &node]() {
-            node.WaitForMessage(std::chrono::seconds(1));
-            unblocked.fetch_add(1);
+        std::mutex cv_mutex;
+        auto       cv       = std::make_shared<std::condition_variable>();
+        auto       received = std::make_shared<std::atomic<bool>>(false);
```
Comment (xkszltl) [inline tests/node_test.cc:37]: include

File: tests/node_test.cc [hunk 12]
```diff
@@ -2,133 +2,250 @@
...
+
+        const int message_value = 0;
+
+        // Different type, never called
+        subscriber.Subscribe<TestMessage<10086>>(channel_subid, [](const auto&) {
+            // Never enters.
+            std::abort();
```
Comment (xkszltl) [inline tests/node_test.cc:44]: 1. Use test asserts or glog
2. include

File: tests/node_test.cc [hunk 13]
```diff
@@ -2,133 +2,250 @@
...
+    auto received = std::make_shared<Received>();
 
     for (std::size_t channel_subid = 0; channel_subid < kNumOfSubChannel; ++channel_subid) {
-        node.Subscribe<TestMessage<0>>(
+        subscriber.Subscribe<TestMessage<0>>(
             channel_subid,
-            [&received, channel_subid](const std::shared_ptr<TestMessage<0>>&) { ++received[0][channel_subid]; });
-        node.Subscribe<TestMessage<1>>(
+            [received, channel_subid, cv](const std::shared_ptr<TestMessage<0>>& msg) {
+                received->data_[0][channel_subid].store(msg->value_);
```
Comment (xkszltl) [inline tests/node_test.cc:99]: If you really need atomic inside predictor, there's probably something wrong. Shared data for cv need to be gated by mutex, and the existence of atomic means someone else may forcefully change it during execution without acquiring the mutex first.

File: tests/node_test.cc [hunk 14]
```diff
@@ -2,133 +2,248 @@
...
+        auto subscriber = std::make_unique<Subscriber>();
+        subscriber->SetRunner(runner.get());
+        subscriber->Subscribe<TestMessageType>(
+            channel_subid,
+            [subscriber = subscriber.get(), cv_mutex, cv](const std::shared_ptr<TestMessageType>& message) {
+                std::unique_lock cv_lock(*cv_mutex);
+                subscriber->received_ = message->value_;
+                cv->notify_all();
+            });
+        subscribers.push_back(std::move(subscriber));
```
Comment (xkszltl) [inline tests/node_test.cc:217]: include

**Review by xkszltl — APPROVED** [review]

**Pull request merged**